# Import Dependencies

In [1]:
import openai

# What is OpenAI API?

The OpenAI API refers to the application programming interface provided by OpenAI, a leading artificial intelligence research lab. The OpenAI API allows developers to access and integrate OpenAI's powerful language models into their own applications, products, or services.

One notable model offered through the OpenAI API is GPT (Generative Pre-trained Transformer). GPT models are designed for natural language processing tasks, such as text generation, completion, translation, summarization, and more. The API provides a way for developers to interact with these models and leverage their capabilities without having to train the models from scratch.

Keep in mind that the specifics of the OpenAI API, including its features and offerings, may have evolved since my last update in January 2022. Therefore, it's a good idea to check the official OpenAI website or other reliable sources for the most up-to-date information on the OpenAI API.


# Generate OPEN API Key

In [2]:
Myapi = "sk-"

In [3]:
openai.api_key = Myapi

In [4]:
All_Models = openai.models.list()

In [5]:
list(All_Models)

[Model(id='curie-search-query', created=1651172509, object='model', owned_by='openai-dev'),
 Model(id='babbage-search-query', created=1651172509, object='model', owned_by='openai-dev'),
 Model(id='dall-e-3', created=1698785189, object='model', owned_by='system'),
 Model(id='babbage-search-document', created=1651172510, object='model', owned_by='openai-dev'),
 Model(id='dall-e-2', created=1698798177, object='model', owned_by='system'),
 Model(id='gpt-3.5-turbo-0301', created=1677649963, object='model', owned_by='openai'),
 Model(id='davinci-search-query', created=1651172505, object='model', owned_by='openai-dev'),
 Model(id='tts-1-hd-1106', created=1699053533, object='model', owned_by='system'),
 Model(id='tts-1-hd', created=1699046015, object='model', owned_by='system'),
 Model(id='ada-search-document', created=1651172507, object='model', owned_by='openai-dev'),
 Model(id='ada-code-search-code', created=1651172505, object='model', owned_by='openai-dev'),
 Model(id='davinci-002', create

In [7]:
import pandas as pd

In [8]:
pd.DataFrame(list(All_Models), columns = ["ID","Created", "Object","Owned_By"])

ID                Created           Object  \
0            (id, curie-search-query)  (created, 1651172509)  (object, model)   
1          (id, babbage-search-query)  (created, 1651172509)  (object, model)   
2                      (id, dall-e-3)  (created, 1698785189)  (object, model)   
3       (id, babbage-search-document)  (created, 1651172510)  (object, model)   
4                      (id, dall-e-2)  (created, 1698798177)  (object, model)   
5            (id, gpt-3.5-turbo-0301)  (created, 1677649963)  (object, model)   
6          (id, davinci-search-query)  (created, 1651172505)  (object, model)   
7                 (id, tts-1-hd-1106)  (created, 1699053533)  (object, model)   
8                      (id, tts-1-hd)  (created, 1699046015)  (object, model)   
9           (id, ada-search-document)  (created, 1651172507)  (object, model)   
10         (id, ada-code-search-code)  (created, 1651172505)  (object, model)   
11                  (id, davinci-002)  (created, 1692634301)  (object, model)   
12                  (id, babbage-002)  (created, 1692634615)  (object, model)   
13      (id, davinci-search-document)  (created, 1651172509)  (object, model)   
14        (id, curie-search-document)  (created, 1651172508)  (object, model)   
15             (id, ada-search-query)  (created, 1651172505)  (object, model)   
16         (id, ada-code-search-text)  (created, 1651172510)  (object, model)   
17       (id, text-embedding-ada-002)  (created, 1671217299)  (object, model)   
18       (id, gpt-3.5-turbo-16k-0613)  (created, 1685474247)  (object, model)   
19                    (id, whisper-1)  (created, 1677532384)  (object, model)   
20            (id, gpt-3.5-turbo-16k)  (created, 1683758102)  (object, model)   
21                (id, gpt-3.5-turbo)  (created, 1677610602)  (object, model)   
22           (id, gpt-3.5-turbo-0613)  (created, 1686587434)  (object, model)   
23           (id, gpt-3.5-turbo-1106)  (created, 1698959748)  (object, model)   
24       (id, gpt-3.5-turbo-instruct)  (created, 1692901427)  (object, model)   
25  (id, gpt-3.5-turbo-instruct-0914)  (created, 1694122472)  (object, model)   
26                   (id, tts-1-1106)  (created, 1699053241)  (object, model)   
27                        (id, tts-1)  (created, 1681940951)  (object, model)   

                       Owned_By  
0        (owned_by, openai-dev)  
1        (owned_by, openai-dev)  
2            (owned_by, system)  
3        (owned_by, openai-dev)  
4            (owned_by, system)  
5            (owned_by, openai)  
6        (owned_by, openai-dev)  
7            (owned_by, system)  
8            (owned_by, system)  
9        (owned_by, openai-dev)  
10       (owned_by, openai-dev)  
11           (owned_by, system)  
12           (owned_by, system)  
13       (owned_by, openai-dev)  
14       (owned_by, openai-dev)  
15       (owned_by, openai-dev)  
16       (owned_by, openai-dev)  
17  (owned_by, openai-internal)  
18           (owned_by, openai)  
19  (owned_by, openai-internal)  
20  (owned_by, openai-internal)  
21           (owned_by, openai)  
22           (owned_by, openai)  
23           (owned_by, system)  
24           (owned_by, system)  
25           (owned_by, system)  
26           (owned_by, system)  
27  (owned_by, openai-internal)

## Code of OpenAI Chat

In [ ]:
from openai import OpenAI
client = OpenAI()

response = client.chat.completions.create(
  model="gpt-3.5-turbo",
  messages=[
    {
      "role": "user",
      "content": "How I can Make Money?"
    }
  ],
  temperature=1,
  max_tokens=256,
  top_p=1,
  frequency_penalty=0,
  presence_penalty=0
)

## This is where you make a request to the OpenAI API for generating a chat completion. Let's break down the parameters:

    model="gpt-3.5-turbo": Specifies the language model to be used, in this case, GPT-3.5-turbo.

    messages: A list of message objects representing the conversation. In this example, there is only one message from the user: "How I can Make Money?".

    temperature=1: A parameter controlling the randomness of the model's output. Higher values like 1 make the output more random, while lower values like 0.2 make it more focused and deterministic.

    max_tokens=256: Specifies the maximum number of tokens (words or characters) in the generated response.

    top_p=1: Top-p (nucleus) sampling probability. Setting it to 1 means the model considers the entire probability distribution for sampling.

    frequency_penalty=0: Controls the penalty for using frequent tokens. A higher value will make the model more creative, while a lower value will make it more conservative.

    presence_penalty=0: Controls the penalty for using new tokens. Similar to frequency_penalty, but for unseen tokens.

# Langchain for Generative AI

Function Calling : Learn how to connect Large Language Models to external tools.

Example invoking multiple function calls in one response

In [ ]:
from openai import OpenAI
import json

client = OpenAI()

# Example dummy function hard coded to return the same weather
# In production, this could be your backend API or an external API
def get_current_weather(location, unit="fahrenheit"):
    """Get the current weather in a given location"""
    if "tokyo" in location.lower():
        return json.dumps({"location": "Tokyo", "temperature": "10", "unit": unit})
    elif "san francisco" in location.lower():
        return json.dumps({"location": "San Francisco", "temperature": "72", "unit": unit})
    elif "paris" in location.lower():
        return json.dumps({"location": "Paris", "temperature": "22", "unit": unit})
    else:
        return json.dumps({"location": location, "temperature": "unknown"})

def run_conversation():
    # Step 1: send the conversation and available functions to the model
    messages = [{"role": "user", "content": "What's the weather like in San Francisco, Tokyo, and Paris?"}]
    tools = [
        {
            "type": "function",
            "function": {
                "name": "get_current_weather",
                "description": "Get the current weather in a given location",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "location": {
                            "type": "string",
                            "description": "The city and state, e.g. San Francisco, CA",
                        },
                        "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
                    },
                    "required": ["location"],
                },
            },
        }
    ]
    response = client.chat.completions.create(
        model="gpt-3.5-turbo-1106",
        messages=messages,
        tools=tools,
        tool_choice="auto",  # auto is default, but we'll be explicit
    )
    response_message = response.choices[0].message
    tool_calls = response_message.tool_calls
    # Step 2: check if the model wanted to call a function
    if tool_calls:
        # Step 3: call the function
        # Note: the JSON response may not always be valid; be sure to handle errors
        available_functions = {
            "get_current_weather": get_current_weather,
        }  # only one function in this example, but you can have multiple
        messages.append(response_message)  # extend conversation with assistant's reply
        # Step 4: send the info for each function call and function response to the model
        for tool_call in tool_calls:
            function_name = tool_call.function.name
            function_to_call = available_functions[function_name]
            function_args = json.loads(tool_call.function.arguments)
            function_response = function_to_call(
                location=function_args.get("location"),
                unit=function_args.get("unit"),
            )
            messages.append(
                {
                    "tool_call_id": tool_call.id,
                    "role": "tool",
                    "name": function_name,
                    "content": function_response,
                }
            )  # extend conversation with function response
        second_response = client.chat.completions.create(
            model="gpt-3.5-turbo-1106",
            messages=messages,
        )  # get a new response from the model where it can see the function response
        return second_response
print(run_conversation())

In [9]:
import langchain

In [10]:
from langchain.llms import OpenAI

In [11]:
client = OpenAI(openai_api_key=Myapi)

C:\Users\Bhavesh\anaconda3\envs\testingopenai\lib\site-packages\langchain_core\_api\deprecation.py:117: LangChainDeprecationWarning: The class `langchain_community.llms.openai.OpenAI` was deprecated in langchain-community 0.0.10 and will be removed in 0.2.0. An updated version of the class exists in the langchain-openai package and should be used instead. To use it run `pip install -U langchain-openai` and import as `from langchain_openai import OpenAI`.
  warn_deprecated(


In [12]:
# Zero short prompting
prompt = "Can you tell me total number of country in Aisa?"

In [13]:
client.predict(prompt)

C:\Users\Bhavesh\anaconda3\envs\testingopenai\lib\site-packages\langchain_core\_api\deprecation.py:117: LangChainDeprecationWarning: The function `predict` was deprecated in LangChain 0.1.7 and will be removed in 0.2.0. Use invoke instead.
  warn_deprecated(


RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}

# Prompt Templates

In [14]:
from langchain.prompts import PromptTemplate

In [15]:
prompt_template = PromptTemplate(
    input_variables=["Country"],
    template="Can you tell me the capital of {Country}?"
)
    

In [16]:
prompt_template.format(Country="Ahmedabad")

'Can you tell me the capital of Ahmedabad?'

In [17]:
prompt_template.format(Country="China")

'Can you tell me the capital of China?'

In [18]:
prompt = PromptTemplate.from_template("What is a good name for a company that makes a {product}")

In [19]:
prompt.format(product="Toys")

'What is a good name for a company that makes a Toys'

# Agents in Langchain

## For extracting a realtime information, We use Serp API
## Using Serp API google search engine
## It will extract information in realtime.

In [28]:
serpapi_key=""

In [29]:
from langchain.agents import AgentType
from langchain.agents import load_tools
from langchain.agents import initialize_agent
from langchain.llms import OpenAI

In [30]:
client = OpenAI(openai_api_key=Myapi)

In [31]:
# !pip install -U langchain-openai

In [32]:
tool = load_tools(["serpapi"], serpapi_api_key=serpapi_key, llm=client)

In [33]:
agent = initialize_agent(tool, client, agent = AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose=True)

C:\Users\Bhavesh\anaconda3\envs\testingopenai\lib\site-packages\langchain_core\_api\deprecation.py:117: LangChainDeprecationWarning: The function `initialize_agent` was deprecated in LangChain 0.1.0 and will be removed in 0.2.0. Use Use new agent constructor methods like create_react_agent, create_json_agent, create_structured_chat_agent, etc. instead.
  warn_deprecated(


In [34]:
agent.run("Can you tell me Who won the Cricket World Cup recently?")

C:\Users\Bhavesh\anaconda3\envs\testingopenai\lib\site-packages\langchain_core\_api\deprecation.py:117: LangChainDeprecationWarning: The function `run` was deprecated in LangChain 0.1.0 and will be removed in 0.2.0. Use invoke instead.
  warn_deprecated(




> Entering new AgentExecutor chain...


RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}

# Notes: It is free Billing OpenAPI that's why this error code 429 shown. 

# Chain

In [ ]:
client

In [20]:
from langchain.prompts import PromptTemplate

In [21]:
prompt = PromptTemplate.from_template("What is a good name for a company that makes {product}")

In [22]:
prompt = prompt.format(product="Wine")

In [23]:
from langchain.chains import LLMChain

In [24]:
LLMChain(llm=client,prompt=prompt)

ValidationError: 1 validation error for LLMChain
prompt
  value is not a valid dict (type=type_error.dict)

In [ ]:
chain.run("Wine").stript()

In [25]:
from langchain.chains import SequentialChain

In [26]:
chain = SequentialChain(
        chains=[name_chain, food_item_chain],
        input_variables=["Cuisine"],
        output_variables-["restaurant_name","menu_item"]
)

                    

SyntaxError: positional argument follows keyword argument (1592107795.py, line 5)

In [27]:
chain({"cuisine":"Indian"})

NameError: name 'chain' is not defined

# Document Loader

In [28]:
# !pip install pypdf

In [29]:
from langchain_community.document_loaders import PyPDFLoader

loader = PyPDFLoader(r"D:/Bhavesh/Documents/Bhavesh Padharia Resume 5-10-2023.pdf")
pages = loader.load_and_split()

In [30]:
pages

[Document(page_content="Bhavesh\nPadharia\nSenior\nData\nScientist\nEmail\n:\npadharia.bhavesh06@gmail.com\n|\nMobile\n:\n77788-58227\n|\nLinkedIn\n:https://www .linkedin.com/in/bhaveshpadharia\n/\nGitHub\n:\nhttps://github.com/padhariabhavesh\nObjective\n:\nHighly\naccomplished\nSenior\nData\nScientist\nwith\n6\nyears\nof\nexperience\nin\ndeveloping\nand\ndeploying\nadvanced\nmachine\nlearning\nmodels,\noptimizing\nprocesses,\nand\ndriving\ndata-driven\nsolutions\nacross\ndiverse\nindustries.\nProven\nexpertise\nin\nmodel\ndevelopment,\ndata\nanalysis,\nand\nproject\nleadership.\nSeeking\nto\nleverage\nmy\nskills\nand\nexperience\nto\ncontribute\nto\na\nforward-thinking\norganization's\ndata\nscience\ninitiatives.\nProfessional\nExperience:\nValue\nChain\nSolutions\n(India)\nPvt.\nLtd\n.\n|\nAhmedabad\n|\nFeb\n23\n-\nPresent\nSenior\nData\nScientist\n●\nModel\nOptimization\nand\nDeployment:\n-\nPlayed\na\ncrucial\nrole\nin\nrefining\nand\ndeploying\nmodel\ncode\nusing\nOSI\nPi\nVision

#  Memory in Langchain

In [7]:
from langchain.prompts import PromptTemplate
from langchain.chains import LLMChain
from langchain.llms import OpenAI

In [8]:
client = OpenAI(openai_api_key=Myapi)

In [10]:
prompt_template_name = PromptTemplate(
    input_variables = ['product'],
    template = " What is a good name for a company  that makes {product}"
)

In [11]:
chain = LLMChain(llm = client, prompt=prompt_template_name)

In [ ]:
chain.run("Colorful Cup")

In [13]:
chain.memory

In [14]:
type(chain.memory)

NoneType

# ConversationBufferMemory

We can attach memory  to remember all previous conversation

In [15]:
from langchain.memory import ConversationBufferMemory

In [16]:
memory = ConversationBufferMemory()

In [17]:
prompt_template_name = PromptTemplate(
    input_variables = ['product'],
    template = " What is a good name for a company  that makes {product}"
)

In [18]:
chain = LLMChain(llm=client, prompt=prompt_template_name,memory=memory)

In [ ]:
chain.run("Wine")

# Conversation Chain


In [19]:
from langchain.chains import ConversationChain

In [20]:
conv = ConversationChain(llm=OpenAI(openai_api_key=Myapi,temperature=0.7)) 

In [21]:
conv.prompt

PromptTemplate(input_variables=['history', 'input'], template='The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.\n\nCurrent conversation:\n{history}\nHuman: {input}\nAI:')

In [24]:
print(conv.prompt.template)

The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
{history}
Human: {input}
AI:


In [ ]:
conv.run("Who won the first world cup?")

In [ ]:
conv.run("Can you tell me how much 5 + 5?")

In [ ]:
conv.run("Can you tell me how much 2*(5+5)?")

In [ ]:
conv.run("Who was the captain of winning team?")

# ConversationBufferWindowMemory

In [25]:
from langchain.memory import ConversationBufferWindowMemory

In [26]:
memory = ConversationBufferWindowMemory( k=1)

In [27]:
conv = ConversationChain(llm=OpenAI(openai_api_key=Myapi,temperature=0.7),memory=memory) 

In [ ]:
conv.run("Who won the first world cup?")